File: frontend/src/api/events.ts
import type { ObserveEvent } from '@/shared/observe'

/**
 * 订阅观测事件流（SSE）。返回退订函数。
 * 事件 kind 由 SSE `event:` 字段给出，data 为 ObserveEvent JSON（不含 kind）。
 */
export function subscribeEvents(onEvent: (kind: string, ev: Partial<ObserveEvent>) => void): () => void {
  const es = new EventSource('/api/v1/events')
  const kinds = ['state', 'activity', 'run']
  for (const kind of kinds) {
    es.addEventListener(kind, (raw) => {
      const msg = raw as MessageEvent
      try {
        onEvent(kind, JSON.parse(msg.data) as Partial<ObserveEvent>)
      } catch {
        /* 忽略坏帧 */
      }
    })
  }
  return () => es.close()
}

File: backend/internal/observe/events.go
// Package observe 提供 v4 的观测聚合与事件总线（ADR-041 §7）。
package observe

import "encoding/json"

// Kind 事件类别。
type Kind string

const (
	KindState    Kind = "state"    // 组件 state 变化
	KindActivity Kind = "activity" // 组件 activity 变化
	KindRun      Kind = "run"      // Run/Step 推进（调和器写入）
)

// Event 观测事件（SSE 载荷）。
type Event struct {
	Seq       uint64          `json:"seq"`
	Kind      Kind            `json:"kind"`
	Component string          `json:"component,omitempty"`
	RunID     string          `json:"runId,omitempty"`
	Data      json.RawMessage `json:"data,omitempty"`
	At        int64           `json:"at"`
}
